File: FSharp/PinetreeChat.Domain/ChatService.fs
namespace PinetreeChat.Domain

module ChatService = 
    open PinetreeChat.Entities
    open Validation
    open Chessie.ErrorHandling   

    type ChatErrorMessage = 
        | ChatNameInvalid of string
        | ChatExists of string
        | MessageInvalid of string

    let getErrorMessage e =
        match e with
        | ChatNameInvalid msg -> msg
        | ChatExists msg -> msg
        | MessageInvalid msg -> msg
    
    let createChat findChatFunction addChatFunction chatName = 
        let notEmpty = validateEmptyString (ChatNameInvalid "Chat name is empty")
        let lengthValid = validateStringLength 1 50 (ChatNameInvalid "Chat name is too long")
        let validateChatName cn = notEmpty cn <* lengthValid cn

        let createChat cn =
            match findChatFunction cn with
            | Some _ -> Bad [ (ChatExists "A chat with this name already exists") ]
            | None ->   
                { Id = 0; Name = cn; Messages = []; Participants = [] }
                |> addChatFunction
                |> ok

        chatName |> validateChatName >>= createChat

    let addMessage (findUserFunction:string -> User option) (addMessageFunction:string -> Message -> (Chat * Message)) (addParticipantFunction:string -> User -> (Chat * User)) chatName text from =
        let notEmtpy = validateEmptyString (MessageInvalid "Message is empty")
        let lengthValid = validateEmptyString (MessageInvalid "Message is too long")
        let validateMessageText tx = notEmtpy tx <* lengthValid tx
        
        let createMessage cn un tx = 
            findUserFunction un
            |> fun(r) -> 
                match r with 
                | Some u -> ok u
                | None -> Bad [( MessageInvalid "No such user" )]
            |> lift (addParticipantFunction cn)
            |> lift (fun (_, u) -> { Id = 0; Order = 0; Text = tx; From = u})
            |> lift (fun (m) -> addMessageFunction cn m)

        text 
        |> validateMessageText 
        >>= createMessage chatName from        

module ChatServiceWithDataAccess =
    open PinetreeChat.DataAccess.ChatRepository
    open PinetreeChat.DataAccess.UserRepository
    open Chessie.ErrorHandling

    let getChat = getChat
    let getChats = getChats
    let createChat = ChatService.createChat getChat addChat
    let addMessage = ChatService.addMessage getUser addMessage addParticipant
    let leaveChat chatName userName = removeParticipant chatName userName |> ok
            

File: CSharp/PinetreeChat.Domain/Services/ChatService.cs
using PinetreeChat.Domain.Services.Exceptions;
using PinetreeChat.Entities;
using PinetreeChat.Interfaces.Repositories;
using System.Collections.Generic;
using System.Linq;

namespace PinetreeChat.Domain.Services
{
    public class ChatService
    {
        private IChatRepository _chatRepo;
        private IUserRepository _userRepo;

        public ChatService(IChatRepository chatRepo, IUserRepository userRepo)
        {
            _chatRepo = chatRepo;
            _userRepo = userRepo;
        }

        public IEnumerable<Chat> GetChats()
        {
            return _chatRepo.GetChats();
        }

        public Chat GetChat(string chatName)
        {
            return _chatRepo.GetChats().SingleOrDefault(c => c.Name == chatName);
        }

        public Chat CreateChat(string chatName)
        {
            if(chatName.Length > 50 || string.IsNullOrWhiteSpace(chatName))
            {
                throw new ChatNameInvalidException(chatName);
            }

            var chats = _chatRepo.GetChats();
            if (chats.Any(c => c.Name.ToLower() == chatName.ToLower()))
            {
                throw new ChatExistsException(chatName);
            }

            var newChat = new Chat(chatName);
            _chatRepo.AddChat(newChat);
            return newChat;
        }

        public Message AddMessage(string chatName, string text, string from)
        {
            if(text.Length > 150 || string.IsNullOrWhiteSpace(text))
            {
                throw new MessageInvalidException(text);
            }

            var chat = _chatRepo.GetChats().SingleOrDefault(c => c.Name == chatName);
            var user = _userRepo.GetUsers().SingleOrDefault(c => c.Username == from);
            if (chat == null)
            {
                return null;
            }

            var message = new Message(user, text);
            _chatRepo.AddMessage(chat, message);

            if (!chat.Participants.Any(p => p.Username == from))
            {
                _chatRepo.AddParticipant(chat, user);
            }

            return message;
        }

        public void LeaveChat(string chatName, string participant)
        {
            var chat = _chatRepo.GetChats().SingleOrDefault(c => c.Name == chatName);
            var user = _userRepo.GetUsers().SingleOrDefault(c => c.Username == participant);
            _chatRepo.RemoveParticipant(chat, user);
        }
    }
}
